content/blog/eng/common-problems-faced-by-php-developers-in-consuming-an-ampq-message.md

```diff
@@ -0,0 +1,298 @@
+languageCode: "en-EN"
+toc: true
+---
+
+When we use RabbitMQ and our project is in PHP, we have to run a Consumer and we could encounter some common and still unresolved problems.
+
+In order to implement RabbitMQ in the project, there are different ready to use packages, especially when we use a framework like Symfony or Laravel. Usually these packages provide the full integration needed for the AMQP protocol, so we can easily configure it and create a Producer and a Consumer; then, we have to create a little script that will supervise the PHP process in order to make sure that all our consumers are running. 
+
+# Consumer
+
+In order to consume a message, we have to run a demon that will listen for new messages, then process each message executing some logic: this is a normal flow in case of blocking requests, and we don't want to block the user request nor loose it.
+
+We want also to keep track of all requests and if one fails, we could retry to process it again or just ignore it if some blocking error occurs.
+
+## The memory problem
+
+A normal PHP script serves a single request, then the script _dies_ naturally as the execution of provided instructions is completed, and the allocated memory is freed. In a PHP consumer we have to execute a long running PHP process: this will cause us to face a memory problem, because PHP is not properly designed to achieve this goal, and this will cause trouble for sure if we don't pay attention to the script code.
+
+Normally, you don't notice any memory issue due to fact that the script runs for few seconds; also, in a long running PHP process, if you try and run a simple script for few minutes it could not use any additional memory at all, but if you use some framework, an ORM or you are in a complex application with many dependencies, the application will eventually allocate some memory (e.g. for caching results), and you could even notice a progressive increase of the memory used by the PHP process due to some memory leaks.
+
+That's not a problem at all in a normal case, where the script ends and the memory is freed normally, but the case is not that obvious for a long running process like the one required by a PHP consumer. It will break, leak and, in some cases, corrupt the memory, and this will make the process crash or, more drastically, fail to run some instructions, as the opcode cache gets changed while the process is running.
+
+If you are interested in more accurate details and information, I suggest reading [this article](http://notjam.es/blog/2014/06/18/the-problems-with-long-running-php.html): it's pretty old, but explains this issue very well.
+
+## Multiple consumers
+
+If you are running a PHP consumer, is very hard to process multiple messages at time because PHP runs in a single thread and also, a proper async/await interface is not well supported or mature. One more time, PHP is not well designed for this scope and this could be a limit in some circumstances, scaling is very hard.
```

Comment:
```suggestion
For a PHP consumer it's very hard to process multiple messages at the same time, because PHP runs in a single thread, and a proper async/await interface is not well supported or mature. Again, PHP is not designed for this scope and this could be a limit in some circumstances: scaling is really hard.
```